### PR TITLE
Initial error refactoring (make errors less HTTP specific)

### PR DESCRIPTION
**Pull request by lightsofapollo**

### CODE REVIEW

**Comment by lightsofapollo** [comment]
@KevinGrandon r?

File: lib/caldav/request/calendar_home.js [hunk 1]
```diff
@@ -72,19 +72,32 @@
...
+          );
+        }
+
+        // we might have both principal.href & unauthenticated
+        if (principal.href) {
+          return callback(null, principal.href);
         }
+
+        callback(
+          new Errors.InvalidEntrypoint('no useful location information found')
```
Comment (KevinGrandon) [inline lib/caldav/request/calendar_home.js:99]: Just a thought.. Maybe you'd save a bit of time on startup work if you instead lazy-init'd the errors? E.g., make the folllowing callable:

```
  new Error('InvalidEntrypoint, 'no useful location information found') 
```

I suppose the current implementation might be a bit more cleaner though.

File: caldav.js [hunk 2]
```diff
@@ -3659,19 +3668,32 @@ function write (chunk) {
...
+            'both current-user-principal and principal-URL are missing'
+          ));
         }
 
+        // per http://tools.ietf.org/html/rfc6638 we get unauthenticated
         if ('unauthenticated' in principal) {
-          callback(new Errors.UnauthenticatedError());
-        } else if (principal.href) {
-          callback(null, principal.href);
-        } else {
```
Comment (KevinGrandon) [inline caldav.js:3672]: Do we no longer need a catch-all case here?

File: lib/caldav/request/calendar_home.js (hunk 1 above)
Comment (lightsofapollo) [inline lib/caldav/request/calendar_home.js:99]: For internal CalDAV use (and use on the main thread) I would like instanceof to work? Make sense?

File: caldav.js (hunk 2 above)
Comment (lightsofapollo) [inline caldav.js:3672]: IIRC this now uses return / fall through model to handle catch-all

**Comment by KevinGrandon** [comment]
@lightsofapollo  Looks good to me! :+1:

**Pull request merged**